src/utils/streams/array_stream.js

```diff
@@ -0,0 +1,55 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { Duplex } from 'stream';
+const NEWLINE_REGEX = /\r?\n/;
+
+export function map(mapfn) {
```

Comment:
We do have https://github.com/elastic/kibana/blob/e7fd20cdf85ae4c3e506fe949bbb66cf18c179fe/src/utils/streams/map_stream.js, which uses promises instead of callbacks for mapping, which would be sufficient/and improvement for several of the usecases. The `xml2js.parseString` case can just be wrapped in a Promise constructor.